### ISSUE TITLE
Using loss as a function of data with `sciml_train`

### ISSUE THREAD

**Issue by ali-ramadhan** [JuliaDiffEq/DiffEqFlux.jl#144]
Currently it seems that `sciml_train`, in contrast to `Flux.train!`, expects the loss to be a function of the parameters and not the data.

I was wondering if it's a good idea to allow users to provide a similar interface to `Flux.train!` where you pass a loss function and data. Might also make mini-batching (#133) easier.

If the loss has to be a function of the parameters deep down then maybe I could write a simple
```julia
sciml_train(loss, θ, data, opt, ...)
```
wrapper that constructs an appropriate loss function and pass it to the current `sciml_train`.

**Comment by ChrisRackauckas** [comment]
I am not sure `data` should be in the training function. I think it can just be enclosed in the loss function, since it is required to be treated as constant?

**Comment by ali-ramadhan** [comment]
Ah hmmm, I was hoping to scale `sciml_train` to 100,000+ training data pairs but if the data is treated as constant, is mini-batching possible?

**Comment by ChrisRackauckas** [comment]
>I guess the loss function could select some subset of the data to compare against but not sure how well that would train. Also seems cleaner to reduce the number of globals floating around.

It would train the same as Flux.train! which also treats the data as constants in terms of differentiation. You can reduce the number of globals just by enclosing the data.
